### ISSUE TITLE
When Papertrail refuses connection you b/c you reached your max

### ISSUE THREAD

**Issue by ddibiase** [kenperkins/winston-papertrail#49]
So I just learned the REALLY hard way that if Papertrail reaches it's max and no longer collects data it will refuse a connection from your servers. That will in turn throw an error with this library:

```
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: connect ECONNREFUSED
    at exports._errnoException (util.js:746:11)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1010:19)
```

It took hours to track down this issue and finally realize what it was. This package should have better error tracking for when a connection with Papertrail can't be made. Took down a production app for an hour. _cries_

**Comment by troy** [comment]
@ddibiase While I'm not the author of this library, I'd like it to be as reliable as any or all of us know how to make it. To that end, what behavior would you recommend for this situation? That is, what should happen when Papertrail is unreachable (whatever the reason - could be that the whole Internet is unreachable)?

**Comment by ddibiase** [comment]
@troy there's already a built in error handler that we could bind to mitigate the problem. I think at a minimum the entire library shouldn't break the whole thread/process if an error callback isn't defined.

**Comment by matteocontrini** [comment]
Well that's the standard behavior when emitting errors in Node.js. 
It's your responsibility to use the library properly.

**Comment by ddibiase** [comment]
@matteocontrini, there are many libraries that handle failures in very graceful ways automatically and there are others that do not. Your argument that a developer *should* use the error handler is accurate but this is really a matter of stability and also convenience in the event that a developer neglects to do so and get burned.

In a package this small, to cause such a catastrophic failure isn't reasonable. In other libraries that are more critical I can see the logic. So take the suggestion for what it is and hopefully other developers using this library head this warning based on this experience.

Cheers!

**Comment by troy** [comment]
@ddibiase re:

> The safest bet is just don't fail like it does.

Up for submitting a PR to do that? It seems like something you've investigated pretty thoroughly and have strong and thoughtfully-formed opinions about. It also seems relatively small to implement. If you're up for taking a stab at what you'd consider an elegant solution -- something that doesn't totally ignore errors but also doesn't fail hard -- I think it might solve this problem.

**Comment by ddibiase** [comment]
@troy it was just an annoying gotcha. I'm happy to contribute I just need to make time, pre-holiday season is always difficult at work.

Thanks though guys, I hope there was some value here.

**Comment by agoldis** [comment]
That's ridiculous, had 8 hours downtime because of the plugin crashing my app.

**Comment by troy** [comment]
@agoldis while I'm not the maintainer.. any chance you're up for sending a PR to fix this? If so, I strongly suspect that it would be accepted and merged.

**Comment by agoldis** [comment]
@troy I've wrote a message to papertrail asking if they support that project, ended up by wrapping adding the transport wrapped in `try catch`. I hope i'll have time to look into the internals, there's also #20 related to that issue/misunderstanding.

**Comment by cymen** [comment]
I ran into this today. I'm in the middle of trying to push the logging down to syslog with papertrail but it was a major bummer. So it's an upstream issue in winston?

**Comment by nvartolomei** [comment]
This one was really sad, the worst thing is that the stack trace is completely useless...

**Comment by JuergenSimon** [comment]
This issue is still alive and kicking. Using winston-papertrail in our node application and it just stopped working today because of this. Took 4 hours to figure out that it was winston-papertrail. The process simply exited with return value 1. There was no stack trace or error message at all. This is a critical bug, because this can happen at any time and effectively sabotages apps. I would expect the error to be handled when the exception handling flag is set without crashing!!

In case it helps: here's our config:

```
var winston = require('winston');
require('winston-papertrail').Papertrail;
var xmlLog = new (winston.Logger)({
  transports: [
    new (winston.transports.File)({
      level: 'info',
      filename: 'xml.log',
      zippedArchive: true,
      maxsize: 10 * 1024 * 1024, // 10MB
      maxFiles: 7,
      tailable: true,
      json: false
    })
    , new (winston.transports.Papertrail) ({
           host: 'logs4.papertrailapp.com',
           port: 52613,
           colorize: true,
           handleExceptions: true
    })
  ]
});

```

**Comment by troy** [comment]
@JuergenSimon what version of `winston-papertrail` and Node do you use? 

Also, does it have an error handler defined, like the `winstonPapertrail.on('error', ..)` stanza [here](https://github.com/kenperkins/winston-papertrail/#usage)?

**Comment by troy** [comment]
Hey all, if you pull `master`, the final fix for this (#61) should now be in the code and will be released to NPM in the next week or two. Even without an error handler defined, connection errors now default to being output to the console but not affecting runtime.

If you pull `master` (or use an eventual `1.0.4` release, when it's out) and see any impact from connection errors, please open a new issue with the transcript and version in use.
